File: services/web-api/src/main/java/com/github/jeffreyjpz/timestamped_for_f1_web_api/web/v1/sessions/dtos/Session.java
package com.github.jeffreyjpz.timestamped_for_f1_web_api.web.v1.sessions.dtos;

import com.fasterxml.jackson.annotation.JsonAlias;
import com.fasterxml.jackson.annotation.JsonProperty;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

@Getter
@RequiredArgsConstructor
public class Session {
    
    @JsonAlias({"circuitKey", "circuit_key"})
    @JsonProperty("circuit_key")
    private final Integer circuitKey;

    @JsonAlias({"circuitShortName", "circuit_short_name"})
    @JsonProperty("circuit_short_name")
    private final String circuitShortName;

    @JsonAlias({"countryCode", "country_code"})
    @JsonProperty("country_code")
    private final String countryCode;

    @JsonAlias({"countryKey", "country_key"})
    @JsonProperty("country_key")
    private final Integer countryKey;

    @JsonAlias({"countryName", "country_name"})
    @JsonProperty("country_name")
    private final String countryName;

    @JsonAlias({"dateEnd", "date_end"})
    @JsonProperty("date_end")
    private final String dateEnd;

    @JsonAlias({"dateStart", "date_start"})
    @JsonProperty("date_start")
    private final String dateStart;

    @JsonAlias({"gmtOffset", "gmt_offset"})
    @JsonProperty("gmt_offset")
    private final String gmtOffset;

    @JsonAlias({"location"})
    @JsonProperty("location")
    private final String location;

    @JsonAlias({"meetingKey", "meeting_key"})
    @JsonProperty("meeting_key")
    private final Integer meetingKey;

    @JsonAlias({"sessionKey", "session_key"})
    @JsonProperty("session_key")
    private final Integer sessionKey;

    @JsonAlias({"sessionName", "session_name"})
    @JsonProperty("session_name")
    private final String sessionName;

    @JsonAlias({"sessionType", "session_type"})
    @JsonProperty("session_type")
    private final String sessionType;

    @JsonAlias({"year"})
    @JsonProperty("year")
    private final Integer year;

}

File: dashboard/src/types/session.ts
import z from "zod";

export const sessionName = z
    .enum([
        "Practice 1",
        "Practice 2",
        "Practice 3",
        "Sprint Shootout", // Sprint Qualifying name pre-2024
        "Sprint Qualifying",
        "Qualifying",
        "Sprint",
        "Race"
    ]);

export const sessionType = z
    .enum([
        "Practice",
        "Qualifying",
        "Race"
    ]);

export const session = z
    .object({
        circuit_key: z.number().int().positive(),
        circuit_short_name: z.string(),
        country_code: z.string().length(3),
        country_key: z.number().int().positive(),
        country_name: z.string(),
        date_start: z.iso.datetime(),
        date_end: z.iso.datetime(),
        gmt_offset: z.string().regex(/^[-+]\d{2}:\d{2}:\d{2}$/),
        location: z.string(),
        meeting_key: z.number().int().positive(),
        session_key: z.number().int().positive(),
        session_name: sessionName,
        session_type: sessionType,
        year: z.number().int().nonnegative(),
    });

export const sessionRead = session
    .pick({
        session_key: true
    });

export const sessionsRead = session
    .extend({
        circuit_key: z.array(z.number().int().positive()),
        circuit_short_name: z.array(z.string()),
        country_code: z.array(z.string().length(3)),
        country_key: z.array(z.number().int().positive()),
        country_name: z.array(z.string()),
        date_start: z.array(z.iso.datetime()),
        date_end: z.array(z.iso.datetime()),
        gmt_offset: z.array(z.string().regex(/^[-+]\d{2}:\d{2}:\d{2}$/)),
        location: z.array(z.string()),
        meeting_key: z.array(z.number().int().positive()),
        session_key: z.array(z.number().int().positive()),
        session_name: z.array(sessionName),
        session_type: z.array(sessionType),
        year: z.array(z.number().int().nonnegative()),
    })
    .partial();

export type SessionName = z.infer<typeof sessionName>;
export type SessionType = z.infer<typeof sessionType>;

export type Session = z.infer<typeof session>;
export type SessionRead = z.infer<typeof sessionRead>;
export type SessionsRead = z.infer<typeof sessionsRead>;